internal/services/serviceprincipals/service_principal_certificate_resource_msgraph.go

```diff
@@ -0,0 +1,164 @@
+	"time"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/manicminer/hamilton/msgraph"
+
+	"github.com/hashicorp/terraform-provider-azuread/internal/clients"
+	helpers "github.com/hashicorp/terraform-provider-azuread/internal/helpers/msgraph"
+	"github.com/hashicorp/terraform-provider-azuread/internal/services/serviceprincipals/parse"
+	"github.com/hashicorp/terraform-provider-azuread/internal/tf"
+)
+
+func servicePrincipalCertificateResourceCreateMsGraph(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*clients.Client).ServicePrincipals.MsClient
+	objectId := d.Get("service_principal_id").(string)
+
+	credential, err := helpers.KeyCredentialForResource(d)
+	if err != nil {
+		attr := ""
+		if kerr, ok := err.(helpers.CredentialError); ok {
+			attr = kerr.Attr()
+		}
+		return tf.ErrorDiagPathF(err, attr, "Generating certificate credentials for service principal with object ID %q", objectId)
+	}
+
+	id := parse.NewCredentialID(objectId, "certificate", *credential.KeyId)
+
+	tf.LockByName(servicePrincipalResourceName, id.ObjectId)
+	defer tf.UnlockByName(servicePrincipalResourceName, id.ObjectId)
+
+	app, status, err := client.Get(ctx, id.ObjectId)
+	if err != nil {
+		if status == http.StatusNotFound {
+			return tf.ErrorDiagPathF(nil, "service_principal_id", "Service principal with object ID %q was not found", id.ObjectId)
+		}
+		return tf.ErrorDiagPathF(err, "service_principal_id", "Retrieving service principal with object ID %q", id.ObjectId)
+	}
+
+	newCredentials := make([]msgraph.KeyCredential, 0)
+	if app.KeyCredentials != nil {
+		for _, cred := range *app.KeyCredentials {
+			if cred.KeyId != nil && *cred.KeyId == *credential.KeyId {
+				return tf.ImportAsExistsDiag("azuread_service_principal_certificate", id.String())
+			}
+			newCredentials = append(newCredentials, cred)
+		}
+	}
+
+	newCredentials = append(newCredentials, *credential)
+
+	properties := msgraph.ServicePrincipal{
+		ID:             &id.ObjectId,
+		KeyCredentials: &newCredentials,
+	}
+	if _, err := client.Update(ctx, properties); err != nil {
+		return tf.ErrorDiagF(err, "Adding certificate for service principal with object ID %q", id.ObjectId)
+	}
+
+	d.SetId(id.String())
+
+	return servicePrincipalCertificateResourceReadMsGraph(ctx, d, meta)
+}
+
+func servicePrincipalCertificateResourceReadMsGraph(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*clients.Client).ServicePrincipals.MsClient
+
+	id, err := parse.CertificateID(d.Id())
+	if err != nil {
+		return tf.ErrorDiagPathF(err, "id", "Parsing certificate credential with ID %q", d.Id())
+	}
+
+	app, status, err := client.Get(ctx, id.ObjectId)
+	if err != nil {
+		if status == http.StatusNotFound {
+			return tf.ErrorDiagPathF(nil, "service_principal_id", "Service principal with object ID %q was not found", id.ObjectId)
```

Comment:
(as.. below?) if the service principals gone, the certificate won't exist - so shouldn't this be removing it from the state?

Reply:
Similarly to passwords, I feel a bit hesitant about removing certificates from state, wouldn't it be safer to have the user run `tf state rm`, or fix their config?

Reply:
nah that seems reasonable, this would show up as 'gone' in the plan - so that's the expected behaviour here